src/connectors/refinement-list/connectRefinementList.js

```diff
@@ -154,126 +154,62 @@ export default function connectRefinementList(renderFn, unmountFn = noop) {
-        // render with previous data from the helper.
-        render({
-          items: lastResultsFromMainSearch,
-          state,
-          createURL,
-          helperSpecializedSearchFacetValues,
-          refine: toggleRefinement,
-          isFromSearch: false,
-          isFirstSearch: false,
-          instantSearchInstance,
-          toggleShowMore, // and yet it will be
-          isShowingMore, // so we need to restore in the state of show more as well
-        });
-      } else {
-        const tags = {
-          highlightPreTag: escapeFacetValues
-            ? TAG_PLACEHOLDER.highlightPreTag
-            : TAG_REPLACEMENT.highlightPreTag,
-          highlightPostTag: escapeFacetValues
-            ? TAG_PLACEHOLDER.highlightPostTag
-            : TAG_REPLACEMENT.highlightPostTag,
-        };
-
-        helper
-          .searchForFacetValues(attribute, query, getLimit(isShowingMore), tags)
-          .then(results => {
-            const facetValues = escapeFacetValues
-              ? escapeFacets(results.facetHits)
-              : results.facetHits;
-
-            const normalizedFacetValues = transformItems(
-              facetValues.map(({ value, ...item }) => ({
-                ...item,
-                value,
-                label: value,
-              }))
-            );
-
-            render({
-              items: normalizedFacetValues,
+    const createSearchForFacetValues = function(helper) {
+      return (
+        state,
+        createURL,
+        instantSearchInstance,
+        isShowingMore
+      ) => query => {
+        if (query === '' && lastResultsFromMainSearch) {
+          // render with previous data from the helper.
+          renderFn({
+            ...this.getWidgetRenderState({
+              results: lastResultsFromMainSearch,
               state,
               createURL,
-              helperSpecializedSearchFacetValues,
-              refine: toggleRefinement,
-              isFromSearch: true,
-              isFirstSearch: false,
               instantSearchInstance,
-              isShowingMore,
-            });
+            }),
+            instantSearchInstance,
           });
-      }
+        } else {
+          const tags = {
+            highlightPreTag: escapeFacetValues
+              ? TAG_PLACEHOLDER.highlightPreTag
+              : TAG_REPLACEMENT.highlightPreTag,
+            highlightPostTag: escapeFacetValues
+              ? TAG_PLACEHOLDER.highlightPostTag
+              : TAG_REPLACEMENT.highlightPostTag,
+          };
+
+          helper
+            .searchForFacetValues(
+              attribute,
+              query,
+              getLimit(isShowingMore),
+              tags
+            )
+            .then(results => {
+              renderFn({
+                ...this.getWidgetRenderState({
+                  results,
+                  state,
+                  createURL,
+                  instantSearchInstance,
+                  isFromSearch: true,
```

Comment:
I'm not sure, but maybe isFromSearch could be part of a second argument to getWidgetRenderState; something like "localRenderState"?

Reply:
I moved it to the upper scope instead, tell me if it is ok for you.